src/Core/DevOps/StaticAnalyze/PHPStan/Rules/NoUpdatesInExecuteQueryRule.php

```diff
@@ -0,0 +1,76 @@
+<?php declare(strict_types=1);
+
+namespace Shopware\Core\DevOps\StaticAnalyze\PHPStan\Rules;
+
+use Doctrine\DBAL\Query\QueryBuilder;
+use PhpParser\Node;
+use PhpParser\Node\Expr\FuncCall;
+use PhpParser\Node\Identifier;
+use PHPStan\Analyser\Scope;
+use PHPStan\Rules\Rule;
+use PHPStan\Rules\RuleErrorBuilder;
+use Shopware\Core\Framework\Log\Package;
+
+/**
+ * @internal
+ *
+ * @implements Rule<FuncCall>
+ */
+#[Package('framework')]
+class NoUpdatesInExecuteQueryRule implements Rule
+{
+    public function getNodeType(): string
+    {
+        return Node\Expr\MethodCall::class;
+    }
+
+    public function processNode(Node $node, Scope $scope): array
+    {
+        if (!($node->name instanceof Identifier)
+            || $node->name->toString() !== 'executeQuery'
+        ) {
+            return [];
+        }
+
+        $errors = [];
+
+        $varType = $scope->getType($node->var);
+        if (\in_array(QueryBuilder::class, $varType->getObjectClassNames(), true)) {
+            $current = $node->var;
+            $hasWriteCall = false;
+
+            while ($current instanceof Node\Expr\MethodCall) {
+                if ($current->name instanceof Identifier) {
```

Comment:
I also tried to write a rule for the QueryBuilder and it is hard to get all cases done :grimacing: 
you can have a look at the test cases I once wrote for the Doctrine PHPStan extension https://github.com/phpstan/phpstan-doctrine/pull/167/files

only case 1 with the chained is working until now :sweat_smile: 

good luck to get it work with every case :see_no_evil:

Reply:
Yes thats true, in my opinion it's not necessary to check all cases here. 
It's not possible to prevent the wrong usage with static analysis either way. You could always build a wrong query in runtime. 

But this rule can prevent yesterday's issue and has also found 21 other occurrences.